File: laocai-bot-spring-boot-starter/src/main/java/indi/dkx/laocai/bot/model/segment/OutgoingImageSegment.java
package indi.dkx.laocai.bot.model.segment;

import com.fasterxml.jackson.annotation.JsonIgnore;
import indi.dkx.laocai.bot.model.enums.ImageSubType;

import java.util.Objects;

/**
 * 出站的图片消息段。
 * <p>
 * 发送时通过 uri 指明图片来源，和入站的临时资源结构不同，协议 type 都是 image。
 */
public final class OutgoingImageSegment implements Segment {

    record Data(String uri, ImageSubType subType, String summary) {
        OutgoingImageSegment toSegment() {
            return new OutgoingImageSegment(this);
        }
    }

    private final Data data;

    private OutgoingImageSegment(Data data) {
        this.data = data;
    }

    public static OutgoingImageSegment of(String uri, ImageSubType subType, String summary) {
        return new OutgoingImageSegment(new Data(Objects.requireNonNull(uri), subType, summary));
    }

    @Override
    public String getType() {
        return "image";
    }

    @Override
    public Data getData() {
        return data;
    }

    @JsonIgnore
    public String getUri() {
        return data.uri();
    }

    @JsonIgnore
    public ImageSubType getSubType() {
        return data.subType();
    }

    @JsonIgnore
    public String getSummary() {
        return data.summary();
    }
}

File: laocai-bot-spring-boot-starter/src/main/kotlin/indi/kyson/laocai/bot/segment/OutgoingImageSegment.kt
package indi.kyson.laocai.bot.segment

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import indi.kyson.laocai.bot.enums.ImageSubType

/**
 * 出站的图片消息段。
 *
 * 发送时通过 uri 指明图片来源，和入站的临时资源结构不同，协议 type 都是 image。
 */
data class OutgoingImageSegment private constructor(override val data: Data) : Segment {

    data class Data(
        val uri: String,
        @JsonProperty("sub_type") val subType: ImageSubType,
        val summary: String?,
    ) {
        fun toSegment(): OutgoingImageSegment = OutgoingImageSegment(this)
    }

    override val type: String
        get() = "image"

    @get:JsonIgnore
    val uri: String
        get() = data.uri

    @get:JsonIgnore
    val subType: ImageSubType
        get() = data.subType

    @get:JsonIgnore
    val summary: String?
        get() = data.summary

    companion object {
        @JvmStatic
        fun of(uri: String, subType: ImageSubType, summary: String?): OutgoingImageSegment =
            OutgoingImageSegment(Data(uri, subType, summary))
    }
}
